**Задание номер 1.**


Для данного задание используется Критерий Согласия Пирсона. 
Вычисляется статистика χ, после чего по таблице определяется соответствующее значение p для данного количества степеней свободы.

$\chi^2$=∑$\frac{(O-E)^2}{E}$

Где O - Observed, наблюдаемое значение

E - expected, ожидаемое значение 

In [4]:
#from IPython.display import Image
from scipy.stats import chisquare
import numpy as np

In [5]:
Obs = np.array([7,12,9,17,10,13,15]) 
#Наблюдаемое количество деревьев на каждом участке. 
#np.array - функция numpy для создания массива

Exp = np.mean(Obs) 
#Ожидаемое среднее количество деревьев на участке при верности нулевой гипотезы
#mean - функция numpy для вычисления среднего
Exp

11.857142857142858

In [6]:
chi_squared = sum((Obs-Exp)**2/Exp)
#вычисление хи-квадрат статистики в соответствии с формулой выше
chi_squared

6.144578313253012

Так как у нас 7 участков, то у нас 7 - 1 = 6 степеней свободы. Соответственно обращаемся к таблице распределения Хи-квадрат к строке с 6 степенями свободы. Видно, что Хи-квадрат 6.144 и 6 степеней значений p-value находится между 0.3 и 0.5. Число сосен варьируется несущественно.

![TABLE](Screenshot.png)

Таблица взята из статьи википедии Chi-square distribution.

In [7]:
chisquare(Obs) 
#проверка с помощью функции chisquare библиотеки scipy.stats
#данная функция принимает на вход массив значений

Power_divergenceResult(statistic=6.144578313253012, pvalue=0.4071909014008217)

Функция подтвердила результат вычислений "вручную" выше

**Ответ**: Число сосен варьируется несущественно

**Задание номер 2.**

Так как число событий досточно велико, то можно использовать приближение биномиального распределения к нормальному и получить Z-статистику, а затем её оценить по таблице. Так как формулировка "Пруды различаются", а не "Один пруд более заражён, чем другой", то берём двухсторонний случай.
Z = (n/N - m/M)/(D1+D2)^0.5

In [8]:
#также подойдёт chi-squared test and fisher's exact test 3-я презентация
n = 20
N = 50
m = 32
M = 50
p = (n+m)/(N+M) # Вероятность успеха
D1 = p*(1-p)/N # Дисперсия
D2 = p*(1-p)/M

In [9]:
p, D1, D2

(0.52, 0.004992, 0.004992)

In [10]:
Z = (n/N - m/M) / (D1+D2)**0.5
Z

-2.401922307076307

Далее сравниваем с таблицей нормального распределения

![TABLE](Screenshot_Z.png)

Получаем p-value чуть менее 0.0082*2=0.0164

In [11]:
#Теперь проверяем вычисления с помощью функции из библиотеки statsmodels
from statsmodels.stats.proportion import proportions_ztest
Z, pval = proportions_ztest([n,m],[N, M], alternative = 'two-sided')
Z, pval #полученные значения совпадают

(-2.401922307076307, 0.016309171877754974)

**Ответ**: На уровне значимости 0.0164 можно утверждать о том, что пруды различаются по заражённости

**Задание номер 3.**

In [12]:
T = 9594
A = 8954
G = 5863
C = 5492
N = 29903
TA = 2377

По сути у нас 9594 T после которых может встретиться A с вероятностью A/N.
То есть 9594 попыток с вероятностью успеха p. 
Попыток много, число успехов велико. Можно использовать нормальное приближение биноминального распределения.

Посчитаем ожидаемое число TA, при предположении независимости появления букв в геноме:
$\mu$ = T\*A/N , а также p и $\sigma$

In [13]:
mu = T*A/N
p = A / N
sigma = (N * p * (1-p)) ** 0.5
mu, p, sigma

(2872.7778483764173, 0.29943483931378123, 79.20139170989613)

Далее смотрим на сколько стандартных отклонений отличается наблюдаемое значение TA от среднего.
Z = (TA-mu)/sigma
Так как смотрим отличие наблюдаемого значения от ожидаемого в любую сторону, то имеем двухсторонний случай

In [14]:
Z = (TA-mu)/sigma
Z

-6.259711321644243

По таблице определяем, что Z = -3.0 было бы достаточно для одностороннего случая, и Z = -3.3 было бы достаточно для двухстороннего случая, чтоб отвергнуть нулевую гипотезу, что нет значимого отличия полученной частоты ТА от ожидаемой

![TABLE](Screenshot_Z.png)

**Ответ**: При пороге значимости $\alpha$ = 0.001 значимое отличие ожидаемой частоты TA от наблюдаемой имеется.

**Задание номер 4.** #можно свести к биноминальному распределению

In [15]:
values = np.array([62.2, 58.7, 59.7, 58.7, 60.0, 62.0, 61.5, 58.2, 61.9, 59.1, 61.3, 60.0, 61.7, 59.6, 61.3, 61.8, 64.3, 62.0, 61.8]) #массив наблюдений

In [16]:
n = len(values)
n #количество наблюдений

19

Я должен отметить, что так как:
а) нигде нет информации о том, что распределение нормальное (Не подойдёт t-test)
б) Число наблюдений всего 19 (не подойдёт Z-test)
в) И даже никак не гарантировано что медиана температуры источника была/есть близка к средней температуре (это мешает использовать парный критерий и критерий знаков)
(Например, значительную часть времени температура источника может быть 50C, но иногда случаются выбросы 120C, что смещает среднее до 61.5С, но не среднее.
Как итог, например, мы можем намерить множество температур от 50 до 60 и один 120С, по результатам теста они почти все окажутся меньше 61.5С, что будет статистически значимо, 
но таким образом мы докажем, что медиана меньше 61.5, а не среднее).

По моему личному мнению, так как об изначальном поведении источника неизвестно ничего, кроме среднего, то следует строить модель на основании температур после землятресения:
А именно, предположить, что распределение новых температур нормальное, проверить на нормальность (например, с помощью критерия Шапиро-Уилка - имеется исследование, по
результатам которого этот критерий, вероятно, обладает наибольшей мощностью
https://www.researchgate.net/publication/267205556_Power_Comparisons_of_Shapiro-Wilk_Kolmogorov-Smirnov_Lilliefors_and_Anderson-Darling_Tests )
После чего сравнить насколько далеко среднее до землятресения отклоняется от высчитанного среднего нормального распределения после землятресения.
При этом используем t-test для n-1=18 степеней свободы, так как количество наблюдений у нас мало. 
Получилось бы что-то в этом роде:

In [17]:
from scipy.stats import shapiro
shapiro(values) #p-несдостаточно, чтоб отвергнуть предположение о нормальности

ShapiroResult(statistic=0.9289318323135376, pvalue=0.1655123084783554)

Далее критерий стьюдента

![TABLE](Screenshot4.png)

In [18]:
from scipy.stats import ttest_1samp
ttest_1samp(values, popmean = 61.5) #pvalue < 0.0803, то есть на уровне значимости 0.0803 мы сможем утверждать, что землятресение повлияло на источник

Ttest_1sampResult(statistic=-1.853427681085353, pvalue=0.08028585110418404)

НО, поскольку вычисления выше - это более сложное построение, вероятно некорректное, то всё-таки предположим, что медиана и среднее источника до землятресения значительно не отличались, и используем критерий знаков.
Смотрим сколько температур выше старого среднего (считаем это успехами), сколько ниже (считаем это неудачами). По нулевой гипотезе вероятность успеха p=0.5
Случай, когда температура равна старому среднему интерпретируем так, как наиболее выгодно для нулевой гипотезы (отбрасывание в данном случае считаю плохой идеей, так как
чисто логически результат равный старому среднему должен уменьшать возможность отринуть нулевую гипотезу. Поэтому считаю лучшим использовать более жёсткий тест).

Используя критерий знаков получаем:

In [19]:
k = sum(values > 61.5) #количество успехов
n = len(values) #всего испытаний
zeros = sum(values == 61.5) #количество случаев, когда температура равна старому среднему
k, n, zeros

(8, 19, 1)

In [20]:
#Интерпретируем температуру равную среднему как успех (таким образом критерий становится строже)
k = k+zeros #9 успехов
# теперь нам нужно посчитать p-value для значений числа успехов 9 или менее. Так как мы смотрим отличие в любую сторону, то это ещё нужно будет умножить на 2.
from scipy.stats import binom
binom.cdf(k=k, n=n, p = 0.5)

0.5000000000000002

Так как при p = 0.5 функция симметрична, а 9 из 19 - даст ровно середину функции распределения, и тест двухсторонний, то p-value 1.00.
Что означает, что данный тест ни на каком уровне значимости не позволит говорить об изменении средней температуры после землетрясения. 
Попробуем использовать парный критерий Вилкоксона, посчитав разницу новых значений и старого среднего (к сожалению, есть одинаковые значения, но в данном случае проблема некритична):

In [21]:
from scipy.stats import wilcoxon
differences = values - 61.5
differences #разница значений температуры и старого среднего

array([ 0.7, -2.8, -1.8, -2.8, -1.5,  0.5,  0. , -3.3,  0.4, -2.4, -0.2,
       -1.5,  0.2, -1.9, -0.2,  0.3,  2.8,  0.5,  0.3])

In [23]:
wilcoxon(x = differences, y = None, alternative = 'two-sided')

WilcoxonResult(statistic=57.0, pvalue=0.21393906073122115)

**Ответ**: При уровне значимости 0.214 или более можем утверждать, что землятресение повлияло на источник 
(хотя, по моему мнению, данных недостаточно, чтоб делать выводы. Был бы рад вашим комментариям по поводу этой задачи)

**Задание номер 5.**

In [29]:
X = np.array([2.49, 1.01, 2.81, 2.60, 3.09, 4.60, 2.50, 1.18, 2.20, 3.12, 2.90, 3.58, 2.89, 4.48, 2.17, 1.26, 2.63, 1.17, 1.72, 1.71, 2.45, 3.75, 3.31, 3.23]) 
#Изделия первого завода

Y = np.array([3.37, 2.12, 3.08, 3.57, 4.79, 1.60, 3.71, 5.48, 3.06, 5.76, 3.86, 4.43, 3.77, 4.59, 2.80, 4.11, 3.70, 2.36, 4.83, 2.33])
#Изделия второго завода

np.mean(X), np.mean(Y) #среднее двух выборок

(2.61875, 3.6660000000000004)

Для данного теста можно использовать ANOVA, которое в данном случае будет эквивалентно двухстороннему t-тесту, так как выборки всего две

In [27]:
from scipy.stats import f_oneway #функция для проведения анализа вариации
f_oneway(X, Y) #данная функция требует подставить в неё массивы, для которых будет проводиться рассчёт. Подставляем 2 массива
#функция считает F-статистику и ассоциированное p-value

F_onewayResult(statistic=11.052098565131523, pvalue=0.0018451142120997227)

**Вывод**: ANOVA указывает на значимое различие средних выборок первого и второго завода (при пороге значимости ~0.002 или более). Болты со второго завода на разрыв в среднем прочнее (при этом у двух выборок предполагается одинаковая дисперсия и происхождение из нормального распределения)

In [30]:
shapiro(X), shapiro(Y) #нулевая гипотеза, что обе выборки происходят из нормальных распределений, критерием Шапиро-Уилка не отклоняется

(ShapiroResult(statistic=0.9654770493507385, pvalue=0.5577559471130371),
 ShapiroResult(statistic=0.9843093752861023, pvalue=0.9770391583442688))